cmd/contour/serve.go

```diff
@@ -275,6 +278,28 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 		FieldLogger: log.WithField("context", "endpointstranslator"),
 	}
 
+	// snapshotCache is used to store the state of what all xDS services should
+	// contain at any given point in time.
+	snapshotCache := cache.NewSnapshotCache(false, cache.IDHash{},
+		log.WithField("context", "XDSServer"))
+
+	// snapshotHandler is used to produce new snapshots when the internal state changes
+	// for any xDS resource.
+	snapshotHandler := &contour.SnapshotHandler{
+		CacheHandler:        ch,
+		EndpointsTranslator: et,
+		SnapshotCache:       snapshotCache,
+	}
+
+	// Pass the snapshotHandler off to the CacheHandler & EndpointsTranslator
+	// so they are able to inform the snapshotHandler that the xDS resources
+	// have changed and should be updated in Envoy.
+	ch.SnapshotHandler = snapshotHandler
+	et.SnapshotHandler = snapshotHandler
```

Comment:
It would be good to try to avoid the circular reference between the cache handler and the snapshots.

Probably what I would try to explore is whether we can take the snapshot in `(*EventHandler) updateDAG()`, and making `(*CacheHandler) OnChange` and free function that creates and returns a new snapshot.

Reply:
Yup I've been working on that from @skriss's suggestion earlier. Will have a new push soon.

Reply:
ok I've done some refactoring on this.